File: day_1/python/day_1.py
from itertools import accumulate

def determine_safe_password(combinations):
    positions = accumulate(
        combinations,
        lambda pos, r: (pos + (int(r[1:]) if r[0] == 'R' else -int(r[1:]))) % 100,
        initial=50
    )
    times_at_zero = sum(1 for pos in positions if pos == 0)
    print(f"Times at zero: {times_at_zero}")

determine_safe_password(open('day_1/combinations.txt').read().splitlines())

# everytime it passes zero, add one to a counter
def determine_safe_password_method_two(combinations):
    position = 50
    times_at_zero = 0
    
    for r in combinations:
        distance = int(r[1:])
        direction = 1 if r[0] == 'R' else -1
        
        # Calculate how many complete loops (times crossing zero)
        laps = distance // 100
        remaining = distance % 100
        
        # Check if the remaining movement crosses zero
        new_position = (position + direction * distance) % 100
        crosses_zero = (
            position != 0 and 
            remaining >= (100 - position if direction == 1 else position)
        )
        
        times_at_zero = times_at_zero + laps + (1 if crosses_zero else 0)
        position = new_position
    
    print(f"Times at zero (method two): {times_at_zero}")
    

determine_safe_password_method_two(open('day_1/combinations.txt').read().splitlines())

File: day_1/day_1.rb
# frozen_string_literal: true

def determine_safe_password(combinations)
  dial_position = 50
  times_at_zero = 0

  combinations.each do |rotation|
    direction = rotation[0]
    distance = rotation[1..].to_i

    dial_position += direction == 'L' ? -distance : distance
    dial_position %= 100

    times_at_zero += 1 if dial_position.zero?
  end

  puts "Times at zero: #{times_at_zero}"
end

determine_safe_password(File.readlines('day_1/combinations.txt', chomp: true))

# rubocop:disable Metrics/AbcSize, Metrics/MethodLength
def determine_safe_password_method_two(combinations)
  dial_position = 50

  times_at_zero = combinations.sum do |rotation|
    distance = rotation[1..].to_i
    step = rotation[0] == 'L' ? -1 : 1

    count = distance.divmod(100).then do |laps, remaining|
      crosses = !dial_position.zero? && remaining >= (step == 1 ? 100 - dial_position : dial_position)
      laps + (crosses ? 1 : 0)
    end

    dial_position = (dial_position + step * distance) % 100
    count
  end

  puts "Times passes zero: #{times_at_zero}"
end
# rubocop:enable Metrics/AbcSize, Metrics/MethodLength

determine_safe_password_method_two(File.readlines('day_1/combinations.txt', chomp: true))
